src/Database/LSMTree/Internal/BlobRef.hs

```diff
@@ -4,185 +4,199 @@
-
-{-# SPECIALISE readBlob ::
-     HasFS IO h
-  -> BlobRef IO (FS.Handle h)
-  -> IO SerialisedBlob #-}
-readBlob ::
+    refs <- VM.new (V.length wrefs)
+    V.iforM_ wrefs $ \i WeakBlobRef {weakBlobRefFile, weakBlobRefSpan} -> do
+      ok <- RC.upgradeWeakReference (blobFileRefCounter weakBlobRefFile)
+      if ok
+        then VM.write refs i StrongBlobRef {
+               strongBlobRefFile = weakBlobRefFile,
+               strongBlobRefSpan = weakBlobRefSpan
+             }
+        else do
+          -- drop refs on the previous ones taken successfully so far
+          VM.mapM_ removeReference (VM.take i refs)
+          throwIO (WeakBlobRefInvalid i)
+    V.unsafeFreeze refs
+
+{-# SPECIALISE removeReference :: StrongBlobRef IO h -> IO () #-}
+removeReference :: (MonadMask m, PrimMonad m) => StrongBlobRef m h -> m ()
+removeReference = BlobFile.removeReference . strongBlobRefFile
+
+{-# INLINE readRawBlobRef #-}
+readRawBlobRef ::
      (MonadThrow m, PrimMonad m)
   => HasFS m h
-  -> BlobRef m (FS.Handle h)
+  -> RawBlobRef m h
+  -> m SerialisedBlob
+readRawBlobRef fs RawBlobRef {rawBlobRefFile, rawBlobRefSpan} =
+    BlobFile.readBlobFile fs rawBlobRefFile rawBlobRefSpan
+
+{-# SPECIALISE readWeakBlobRef :: HasFS IO h -> WeakBlobRef IO h -> IO SerialisedBlob #-}
+readWeakBlobRef ::
+     (MonadMask m, PrimMonad m)
+  => HasFS m h
+  -> WeakBlobRef m h
   -> m SerialisedBlob
-readBlob fs BlobRef {
-              blobRefFile,
-              blobRefSpan = BlobSpan {blobSpanOffset, blobSpanSize}
-            } = do
-    let off = FS.AbsOffset blobSpanOffset
-        len :: Int
-        len = fromIntegral blobSpanSize
-    mba <- P.newPinnedByteArray len
-    _ <- FS.hGetBufExactlyAt fs blobRefFile mba 0
-                             (fromIntegral len :: FS.ByteCount) off
-    ba <- P.unsafeFreezeByteArray mba
-    let !rb = RB.fromByteArray 0 len ba
-    return (SerialisedBlob rb)
-
-readBlobIOOp ::
-     P.MutableByteArray s -> Int
-  -> BlobRef m (FS.Handle h)
-  -> FS.IOOp s h
-readBlobIOOp buf bufoff
-             BlobRef {
-               blobRefFile,
-               blobRefSpan = BlobSpan {blobSpanOffset, blobSpanSize}
-             } =
-    FS.IOOpRead
-      blobRefFile
-      (fromIntegral blobSpanOffset :: FS.FileOffset)
-      buf (FS.BufferOffset bufoff)
-      (fromIntegral blobSpanSize :: FS.ByteCount)
+readWeakBlobRef fs wref =
+    bracket (deRefWeakBlobRef wref) removeReference $
+      \StrongBlobRef {strongBlobRefFile, strongBlobRefSpan} ->
+        BlobFile.readBlobFile fs strongBlobRefFile strongBlobRefSpan
+
+{-# SPECIALISE readWeakBlobRefs :: HasBlockIO IO h -> V.Vector (WeakBlobRef IO h) -> IO (V.Vector SerialisedBlob) #-}
+readWeakBlobRefs ::
+     (MonadMask m, PrimMonad m)
+  => HasBlockIO m h
+  -> V.Vector (WeakBlobRef m h)
+  -> m (V.Vector SerialisedBlob)
+readWeakBlobRefs hbio wrefs =
+    bracket (deRefWeakBlobRefs wrefs) (V.mapM_ removeReference) $ \refs -> do
```

Comment:
And this is `withWeakBlobRefs`, which is also removed later. Might be okay to keep that function

Reply:
Yes, inlined here because it's the only use. We can always add it back if we need it. We'd also export `StrongBlobRef`, which does not need to be exported.